apps/dotcom/sync-worker/src/replicator/ChangeCollator.ts

```diff
@@ -0,0 +1,396 @@
+		// when actually delivering the changes.
+		return true
+	}
+
+	/**
+	 * Find all users who should receive changes for any of the given topics.
+	 * Uses recursive SQL to traverse the subscription graph and find all
+	 * users connected to these topics through any number of hops.
+	 *
+	 * Example: user:alice -> group:dev -> file:doc1
+	 * When file:doc1 changes, alice gets notified even though she's not directly subscribed.
+	 */
+	private getSubscribersToTopics(topics: Topic[]): string[] {
+		if (topics.length === 0) return []
+
+		// Extract user topics directly without SQL query
+		const directUsers = topics
+			.filter((topic) => topic.startsWith('user:'))
+			.map((topic) => topic.replace('user:', ''))
+
+		// Get non-user topics that need recursive lookup
+		const nonUserTopics = topics.filter((topic) => !topic.startsWith('user:'))
+
+		if (nonUserTopics.length === 0) {
+			return directUsers
+		}
+
+		// Use recursive CTE to find all users that subscribe to any of these topics
+		// Build placeholders for the IN clause
+		const placeholders = nonUserTopics.map(() => '?').join(',')
+
+		const result = this.sqlite
+			.exec<{ fromTopic: string }>(
+				`
+			WITH RECURSIVE subscribers AS (
+				-- Direct subscriptions to any of the target topics
+				SELECT fromTopic FROM topic_subscription WHERE toTopic IN (${placeholders})
+				UNION
+				-- Transitive subscriptions (users -> groups -> topics)
+				SELECT ts.fromTopic 
+				FROM topic_subscription ts
+				JOIN subscribers s ON ts.toTopic = s.fromTopic
+			)
+			SELECT DISTINCT fromTopic FROM subscribers WHERE fromTopic LIKE 'user:%'
+		`,
+				...nonUserTopics
+			)
+			.toArray()
+
+		const recursiveUsers = result.map((row) => row.fromTopic.replace('user:', ''))
+
+		// Combine and deduplicate all users
+		return [...new Set([...directUsers, ...recursiveUsers])]
+	}
+
+	/**
+	 * Queue a change for delivery to all users who subscribe to any of the given topics.
+	 * Uses the subscription graph to fan out the change to all relevant users.
+	 * Ensures each user gets the change only once even if they're subscribed to multiple topics.
+	 */
+	addChangeForTopics(topics: Topic[], change: ZReplicationChange) {
+		// Get all unique users across all topics in a single SQL query
+		const userIds = this.getSubscribersToTopics(topics)
+
+		// Add the change once per unique user
+		for (const userId of userIds) {
+			let changes = this.changes.get(userId)
+			if (!changes) {
+				changes = []
+				this.changes.set(userId, changes)
+			}
+			changes.push(change)
+		}
+	}
+
+	handleEvent(change: ChangeV2) {
+		// we update subscriptions during both normal operation and catch up
+		const { command, table } = change.event
+
+		if (!this.isCatchUp) {
+			const effects = getEffects(change)
+			if (effects) {
+				this.effects.push(...effects)
+			}
+		}
+
+		// Check if any of the topics have listeners
+		const hasAnyListener = this.hasListenerForTopics(change.topics)
```

Comment:
Got thrown off a bit by this as I thought this is always `true` since it lives inside `LiveChangeCollator`. I guess I'd probably pull this method up to an abstract class `ChangeCollator`, then have two proper subclasses of that. Also fine if you keep it like is though.